Add audio data provider and a simplified DeepSpeech2 model configuration.

xinghai-sun opened the pull request:
resolved [issue 2226](https://github.com/PaddlePaddle/Paddle/issues/2226)  
resolved [issue 2231](https://github.com/PaddlePaddle/Paddle/issues/2231)

xinghai-sun committed:
Add deep_speech_2 folder.

xinghai-sun committed:
Add librispeech dataset, audio data provider and simplfied DeepSpeech2 model configuration.
Bug exists when run training.

xinghai-sun committed:
Add infererence and add SortaGrad for only first pass.

xinghai-sun committed:
Add function docs.

xinghai-sun committed:
Update some parameters and comments.

xinghai-sun commented on deep_speech_2/train.py:
```diff
@@ -0,0 +1,128 @@
+            feeding=feeding)
+        args.num_passes -= 1
+    # other passes without sortagrad
+    trainer.train(
```
If `args.use_sortagrad` is true, `trainer.train` will be called twice. However, the second `trainer.train` call will get stuck (no progress, no error).  `trainer.train` does not support multiple function calls ?

xinghai-sun replied:
I think more elaborate controlling interfaces of training process need to be exposed to users for more flexible training flow control. E.g. in this DS2 case, training data needs to be changed during training. In other cases, parts of model needs freezing for a while or being trained alternatively (e.g. GAN).

lcy-seso commented on deep_speech_2/model.py:
```diff
@@ -0,0 +1,155 @@
+
+    forward = paddle.layer.recurrent_group(
+        step=__simple_rnn_step__, input=input)
+    return forward
```
This has been fixed. Need to be updated.

xinghai-sun replied:
Done.

lcy-seso commented on deep_speech_2/train.py:
```diff
@@ -0,0 +1,128 @@
+        if isinstance(event, paddle.event.EndPass):
+            result = trainer.test(reader=test_batch_reader, feeding=feeding)
+            print "Pass: %d, TestCost: %s" % (event.pass_id, result.cost)
+            with gzip.open("params.tar.gz", 'w') as f:
```
Save the trained model according to the pass number, otherwise, the later saved model overwrites the former ones.

xinghai-sun replied:
Since overfitting rarely happens in DS2, it is not necessary to save multiple models with the pass index. Currently the latest one will be enough. Might add it in the future if necessary.

lcy-seso commented on deep_speech_2/model.py:
```diff
@@ -0,0 +1,155 @@
+        size=dict_size + 1,
+        blank=dict_size,
+        norm_by_times=True)
+    # max decoder
```
If `max_id` is not needed in training, I think it should be put into a  testing branch.

xinghai-sun replied:
Will refactor this part later, together with beam search decoder.

lcy-seso commented on deep_speech_2/infer.py:
```diff
@@ -0,0 +1,109 @@
+        rnn_size=args.rnn_layer_size)
+
+    # load parameters
+    parameters = paddle.parameters.Parameters.from_tar(
```
Save / Load models according to the pass index.

xinghai-sun replied:
Since overfitting rarely happens in DS2, it is not necessary to save multiple models with the pass index. Currently the latest one will be enough. Might add it in the future if necessary.

lcy-seso reviewed (changes requested):
almost LGTM.

xinghai-sun committed:
Refactor data utils into a class and add feature normalization.

qingqing01 commented on deep_speech_2/librispeech.py:
```diff
@@ -0,0 +1,106 @@
+
+URL_TEST = "http://www.openslr.org/resources/12/test-clean.tar.gz"
+URL_DEV = "http://www.openslr.org/resources/12/dev-clean.tar.gz"
+URL_TRAIN = "http://www.openslr.org/resources/12/train-clean-100.tar.gz"
```
Add `MD5` to check data.

xinghai-sun replied:
Done.

qingqing01 commented on deep_speech_2/audio_data_utils.py:
```diff
@@ -0,0 +1,383 @@
+
+class DataGenerator(object):
+    """
+    DataGenerator provides basic audio data preprocessing pipeline, and offer
```
offer -> offers

xinghai-sun replied:
Done.

qingqing01 commented on deep_speech_2/librispeech.py:
```diff
@@ -0,0 +1,106 @@
+    return target_dir
+
+
+def create_manifest(data_dir, manifest_path):
```
What manifest mean?  Add some comments.

xinghai-sun replied:
Done.

qingqing01 commented on deep_speech_2/model.py:
```diff
@@ -0,0 +1,155 @@
+    fc = paddle.layer.fc(
+        input=rnn_group_output,
+        size=dict_size + 1,
+        act=paddle.activation.Linear(),
```
Note,  the activation should be `softmax` in inference mode.

xinghai-sun replied:
Current code only contains the Best Path Decoder, which do not require a `softmax` activation.

qingqing01 commented on deep_speech_2/requirements.sh:
```diff
@@ -0,0 +1,5 @@
+pip install soundfile
+
+# For Ubuntu only
+apt-get install libsndfile1
```
Need to explain in the document.

pkuyym replied:
python部分可以改成提供一个 requires.txt

xinghai-sun replied:
Done.

pkuyym commented on deep_speech_2/audio_data_utils.py:
```diff
@@ -0,0 +1,383 @@
+                 stride_ms=10.0,
+                 window_ms=20.0,
+                 max_frequency=None):
+        self.__max_duration__ = max_duration
```
Python里面最好不要使用__XX__来定义自己的函数和变量，因为__init__、__del__等是python内置的命名方式，下同。如果定义私有的函数或变量只在前面加下划线就可以，印象中__XX是私有的，子类无法访问，_XX也是私有的但是子类可访问

xinghai-sun replied:
目前参考paddle相关python规范如此，建议和大家讨论之后再改？

pkuyym commented on deep_speech_2/audio_data_utils.py:
```diff
@@ -0,0 +1,383 @@
+        self.__stride_ms__ = stride_ms
+        self.__window_ms__ = window_ms
+        self.__max_frequency__ = max_frequency
+        self.__random__ = random.Random(RANDOM_SEED)
```
是否考虑开放接口给用户，让其指定随机种子？

xinghai-sun replied:
RANDOM_SEED 不重要，不建议开放。

pkuyym commented on deep_speech_2/model.py:
```diff
@@ -0,0 +1,155 @@
+        norm_by_times=True)
+    # max decoder
+    max_id = paddle.layer.max_id(input=fc)
+    return cost, max_id
```
可以根据训练或者预测分别返回cost或max_id，如果是预测的话，cost应该是非必须的吧

xinghai-sun replied:
Will refactor this part later when merging with beam search decoder.

pkuyym reviewed (commented):
Almost LGTM.

kuke commented on deep_speech_2/train.py:
```diff
@@ -0,0 +1,141 @@
+    "--use_gpu", default=True, type=bool, help="Use gpu or not.")
+parser.add_argument(
+    "--use_sortagrad", default=False, type=bool, help="Use sortagrad or not.")
+parser.add_argument(
```
第20行已经定义了"--trainer"，help信息相同，是否重复定义？

xinghai-sun replied:
Yes. Removed.

kuke reviewed (commented):
almost LGTM

xinghai-sun reviewed (commented):
Done. Thanks for the review!

xinghai-sun reviewed (commented):
Thanks for the review.

xinghai-sun committed:
1. Fix incorrect decoder result printing.
2. Fix incorrect batch-norm usage in RNN.
3. Fix overlapping train/dev/test manfests.
4. Update README.md and requirements.txt.
5. Expose more arguments to users in argparser.
6. Update all other details.

xinghai-sun commented:
Now the model can run smoothly, with a good convergence and reasonable decoding results.

lcy-seso reviewed (approved):
LGTM
